lib/dynflow/middleware/common/singleton.rb

```diff
@@ -0,0 +1,24 @@
+module Dynflow
+  module Middleware::Common
+    class Singleton < Middleware
+      def plan(*args)
+        action.singleton_lock!
+        pass(*args)
+        unless action.respond_to?(:run) || action.respond_to?(:finalize)
+          action.singleton_unlock!
+        end
+      end
+
+      def run(*args)
+        action.singleton_lock! unless action.holds_singleton_lock?
+        pass(*args)
+        action.singleton_unlock! unless action.respond_to?(:finalize)
```

Comment:
Few questions:

1. what happens when the action goes to suspended state during the run phase?
2. what happens if the action fails? What if the user skips the action on failure? Would the lock stay active?
3. what happens to the lock when the user deletes the execution plan? Shouldn't we have a mechanism to invalidate the locks if the holding execution plan is missing? Also probably deleting the lock on execution plan deletion would be a good idea
4. what should happen on the task pause? Should it unlock or keep the lock. I would vote for unlocking, to let spawn a new task instead of resuming.

Wouldn't be a solution to avoid this questions to move the releasing of the `SingletonActionLock` to the execution plan, and hook it on the change of the execution plan state?

Reply:
1) action suspends itself and keeps the lock (we even have a test for that), when it receives an event it asserts it still holds the lock and continues
2) The execution plan goes to paused-error, and may (in case of any unhandled exception) or may not (in case `#error!` was used) keep the lock. By default it doesn't do anything with the lock when skipping.
3) the locks get released when the execution plan is invalidate. This is probably not enough and a more robust solution is needed to handle all the other cases.
4) Originally I thought the task should keep the lock until it goes to stopped, but unlocking would probably be better. Letting a new task spawn sounds good

Moving the unlocking to execution plan would be probably nicer, I'll try looking into it